app/controllers/auth_controller.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+require 'saml/settings'
+
+class AuthController < ApplicationController
+  skip_before_action :authenticate_user!
+  skip_before_action :verify_authenticity_token
+
+  def new
+    request = OneLogin::RubySaml::Authrequest.new
+    redirect_to(request.create(saml_settings))
+  end
+
+  def callback
+    response = OneLogin::RubySaml::Response.new(params[:SAMLResponse],
+                                                settings: saml_settings)
+    if response.is_valid?
+      user = User.from_saml_callback(response.attributes)
+      if user
+        sign_in_and_redirect user
+      else
+        flash.alert = 'User does not have a valid GIDS system account. ' \
+                      'Contact the system admin for further assistance.'
+        redirect_to root_url + '?auth=fail'
+      end
+    else
+      flash.alert = "Invalid SAML response. Errors: #{response.errors.join("\n")}"
```

Comment:
I don't think this is the alert we want to display to users. Should be a nicer user-facing message.